README
Data Specification
------------------
1.raw data with 14/16 columns(fields).
#   # FIELD  14col,16col
    # IMEI      0, 3
    # IMSI      1, 4
    # UA        2, 5
    # MCC       3, 6
    # MNC       4, 7
    # LAC       5, 8
    # CI        6, 9
    # rss       7, 10
    # lat       8, 11
    # lon       9, 12
    # h         10,13
    # wlanmacs  11,14
    # wlanrsss  12,15
    # Time      13,2
2. columns definition for reqreterr data file and addcols structure.
# reqreterr format:    whole           addcols
# macs                  0,                    
# rsss                  1,      
# ref(lat,lon)          2,3,    
# cpp(lat,lon,err)      4,5,6,          
# ecpp                  7,              0,      
# ee_cpp                8,              1,      
# py(lat,lon,err)       9,10,11,        2,3,4,    
# epy                   12,             5,
# ee_py                 13,             6,
# e_cpp_py              14,             7,
# ee_cpp_py             15,             8,
# google(lat,lon,err)   16,17,18,       9,10,11
# err_google            19,             12,
# ee_google             20,             13,

Known Issues
------------
1. visAPs = [ ['00:19:E0:E3:85:D0','-92'] ], cid = 1304, timestamp=20101013-152014 
2. visAPs = [ ['00:17:7B:0F:85:F0','-68'] ], cid = 1304, timestamp=20101013-095222
3. sum_rss = [ 0, 0, 0, ... ], when issue 1 happens.

Quick Start
------------
1.offline
    1) fingerprints calibration.
        a) ./tool/bt-dev-con && xgps &
        b) ./run -n <start_spid> -r <repeat_times>
        c) ./offline.py -c dat/<time-spid>.rmp
        d) ./offline.py -u 1
    2) Dumbtest for stability.
        ./run -d
    3) Compile all *.py to *.pyc.
        ./run -m

2.online
    1) Pure fingerprinting.
    sudo ./online.py -v
    2) Accuracy evaluation(GPS available).
    sudo ./test.py -v


SP id assignment
----------------
    borq: floor 8: 1-12
    floor 1-8: 100-899
    dacheng floor 12: 1200-1299,  e.g. 1203:door1203.
    home yjl: 90-99
    home neighborhood: 30-89


CHANGELOG
2010-01-06 
    1) Parameters naming changed (-t,-s).
    2) Mac list intersection solved (test/intersect.py).
2010-01-08 
    1) Add --no-dump parameter with no log file writting.
    2) Mod MAC:RSS output format under verbose mode.
2010-01-13
    1) Add online location fingerprinting routine.
2010-01-14
    1) Mod returned spid,rss structure, and the final computation of the KNN location.
    2) Add run test script
    3) Mod online location result output format.
    4) Fix ary_kmin array reshape error when there are not enough KNN fingerprints.
    5) Opt WLAN scanned AP number less than INTERSIZE.
2010-01-16
    1) Add multiple fake GPS mode(home/cmri,etc) handling in online phase.
    2) Add concise address information support.
2010-01-17
    1) Update issue 5: interset solution error when strongest AP is not in radio map.
    2) Fix issue 5: break -> continue.
    3) Mod fake WLAN data config.py.
2010-01-20
    1) Fix MAC:RSS descending order sorted based on RSS values for Radio map file generation.
2010-01-21
    1) Mod offline RadioMap(), only certain number(INTERSIZE) of max-rss APs left in radio map fingerprints.
    2) Fix logical procedure bug for maxmacs generation, introduced by REV<e65c339eeb>.
    3) Fix appending error for rss_scan_dist, introduced by using maxmacs instead of mac_inter.
    4) Fix onopts bug in run script, introduced by REV<5bc8b654fd>.
2010-01-22
    1) Add offline clustering heuristics.
2010-01-24
    1) Clustering heuristics implementation: Cluster(rmpfile). The circumstance that fingerprint's APs included in member of sets_keyaps list should be considered.
2010-01-25
    1) KML interface agile development for sampling and location point demonstration.
2010-01-26
    1) KML related data structure specification and function encapsulation.
    2) Clear up clustering heuristics.
    3) mod  KML __main__ to contain testing data in dat/ap.dat, which should be deleted.
2010-01-27
    1) Data organization for two designated sql tables(cid_aps,cfps) in clustering routine.
2010-01-28
    1) Add MySQLdb interface for clustered fingerprints uploading into database tables.
2010-01-29
    1) Opt code trimming and efficiency of maxmacs and maxrsss construction routine with vectorized operation.
    2) Add online db operation and immature data processing for location fingerprinting.
2010-02-01
    1) bare bone implementation of online processing.
2010-02-02
    1) Fix issue 7 by re-arranging RSSs according to key MACs in table 'cfps'.
2010-02-03
    1) Merge opt online list atof operation for keyrsss and todo list updating.
2010-02-04
    1) Update tbl data after offline calibration, filtering, clustering and sql importing.
    2) Unclustered fingerprints spid printing in offline procedure.
    3) Coherent offline processing support in run -n <start_spid> argument.
2010-02-05
    1) Fix issue 10 for maxmacs subset-searching in fingerprints.
    2) Fix goon assignment error in run script.
2010-02-07
    1) Fix bug when wlan len == 0.
    2) Fix bug when subset searching fails.
    3) Mod trimming imported file handling related parameters.
2010-02-08
    1) Add *nix kernel syscall method for WLAN AP scanning.
2010-02-09
    1) Fix exception handling bug in scanWLAN_OS() of WLAN module.
    2) Exception handling for GPS module.
2010-02-10
    1) Fix RSS calculation error(minus 256) in scanWLAN_OS() of WLAN module.
	2) Update TODO list.
2010-02-11
    1) Fix issue 11, fast fix when ONLY 1 cluster selected in 'cidaps' has ONLY 1 spid selected in 'cfps'.
    2) Clean up online file related code, backward support terminated.
    3) Opt data organization for DB config: store data in dicts.
2010-02-12
    1) Add db_config import in offline routine as what online does.
2010-02-20
    1) Fix robustness bug for optimizing intersection size by sorting and indexing intersection size list of all clusters, instead of using defected loop.
    2) Fix issue 14 when interpart_offline=True, however interpart_online=False.
    3) Fix minor print formating bug when keycfps has more than 1 element.
2010-02-21
    1) Add KML generation routine for pinpointing offline sampling points.
2010-02-25
    1) Fix issue 15 for 'interpart_online' is true with more than one overlapping macs.
    2) Add exception handling for root permission error(errno.EPERM).
2010-02-26
    1) Add dumb test in run script with support for redirecting stdout, stderr to different logs.
    2) Mod fake WLAN_FAKE id.
2010-02-27
    1) Add exception handling for key looking up in dict WLAN_FAKE.
2010-02-28
    1) Mod solution of maxrsss thru array.astype(int) instead of list comprehension.
    2) Opt DKNN location fix process.
2010-03-01
	1) Fix CLUSTERKEYSIZE restriction on the number of offline visible APs.
    2) Add functionalization for getWLAN(). Combine maxmacs/maxrsss into one array:wlan.
	3) Fix date/timestamp fusion, fakegps init value and option error in run.
2010-03-02
	1) Add test module for verification of algo performance.
	2) Add fixPos() with uniformed returned loc, peel off getWLAN(), and rm global vars in online module.
	3) Add ExtGPS data source connection script: bt-dev-con.
	4) Add pyc compileall option '-m' for run.
	5) Update and Fix issue 16.
2010-03-04
    1) Add distance calculation between refloc and fixloc in test module.
	2) reAdd CLUSTERKEYSIZE limit on number of APs for radio map generation. 
    3) Fix bt-dev-con grep bug.
2010-03-06
    1) Add 'title' field in KML input data mandatory section.
    2) Add Google Map js interface wrapper.
2010-03-08
    1) Mod Map module.
2010-03-09

config.py
#-*- encoding=utf8 -*-
DATPATH = 'dat/'
LOCPATH = DATPATH+ 'loc/'
RAWSUFFIX = '.raw'
RMPSUFFIX = '.rmp'
LOCSUFFIX = '.loc'
#INTERSIZE = 10
CLUSTERKEYSIZE = 4
KNN = 4
KWIN = 1.25
RADIUS = 6372797 #meter

# FPP-WPP rawdata sync related.
ftpcfgs = { 
  'fpp_neu': {
       'user' : 'fpp',
     'passwd' : 'fpp',
          'ip': '192.168.109.58',
        'port': 21,
        'path': 'fpp/data_sync/wpp',
     'timeout': 20,
    'localdir': '/root/wpp/src/wlan-pos/dat/sync'},
  'local': {
       'user' : 'alexy',
     'passwd' : 'yan714257',
          'ip': 'localhost',
        'port': 21,
        'path': 'tmp/wpp/ftp',
     'timeout': 10,
    'localdir': '/home/alexy/tmp/wpp/local'},
}
FTPCFG=ftpcfgs['fpp_neu']
mailcfg = {
      'from' : 'xiaotian.yan@gmail.com',
       'to'  : '13811310350@139.com',
    'userpwd': ('xiaotian.yan','yan714257'),
}
errmsg = { 'db' : """
TABLE: [%s]
OPERATION: [%s]
DETAILS: %s

--
WPP@%s
%s""",
}
# DB related configuration.
#DB_ONLINE = 'local_pg'
DB_ONLINE = '192.168.109.54'
DB_OFFLINE = ( '192.168.109.54', ) # Incr clustering.
DB_UPLOAD = ( '192.168.109.54', )
dsn_local_ora = "yxt/yxt@localhost:1521/XE"
dsn_vance_ora = "mwlan/mwlan_pw@192.168.35.202/wlandb"
dsn_local_pg  = "host=localhost dbname=wppdb user=wpp password=wpp port=5432"
dsn_vance_pg_mic = "host=192.168.19.132 dbname=wpp user=mwlan password=mwlan_pw port=5432"
dsn_vance_pg = "host=192.168.109.49 dbname=wppdb user=mwlan password=mwlan_pw port=5432"
dsn_moto_pg = "host=192.168.109.54 dbname=wppdb user=wpp password=wpp port=5432"
dbtype_ora = 'oracle' 
dbtype_pg  = 'postgresql'
dbtype_my  = 'mysql'
dbsvrs = {'192.168.19.132':{
            'dsn':dsn_vance_pg_mic,
            'dbtype':dbtype_pg,
           },
          '192.168.109.54':{
            'dsn':dsn_moto_pg,
            'dbtype':dbtype_pg,
           },
          '192.168.109.49':{
            'dsn':dsn_vance_pg,
            'dbtype':dbtype_pg,
           },
          '192.168.35.202':{
            'dsn':dsn_vance_ora,
            'dbtype':dbtype_ora,
           },
          'local_ora':{
            'dsn':dsn_local_ora,
            'dbtype':dbtype_ora,
           },
          'local_pg':{
            'dsn':dsn_local_pg,
            'dbtype':dbtype_pg,
           },
        }
db_config_my = {
            'hostname' : 'localhost',
            'username' : 'pos',
            'password' : 'pos',
              'dbname' : 'wlanpos' }
# SQL table related data structs.
wpp_tables_my = { 'cidaps' : 'cidaps', 
                    'cfps' : 'cfps'}
#wpp_tables_my = ( 'cidaps', 'cfps' )
tbl_field_my = { 'cidaps':'(cid, keyaps, seq)',
                   'cfps':'(cid, lat, lon, height, rsss, cfps_time)' }
tbl_forms_my = {'cidaps':""" (
                     cid SMALLINT NOT NULL, 
                  keyaps VARCHAR(1024),
                     seq SMALLINT,
                   INDEX icid (cid)
                )""", 
                'cfps':""" (
                     cid SMALLINT NOT NULL,
                     lat DOUBLE(9,6),
                     lon DOUBLE(9,6),
                  height DOUBLE(5,1),
                    rsss VARCHAR(255),
               cfps_time VARCHAR(20),
                   INDEX icid (cid)
                )""" }
# { table_name: table_instance }
wpp_tables = { 'wpp_clusteridaps':'wpp_clusteridaps',
                       'wpp_cfps':'wpp_cfps',
                 'wpp_uprecsinfo':'wpp_uprecsinfo',
                  'wpp_uprecsver':'wpp_uprecsver',
               'wpp_uprecs_noloc':'wpp_uprecs_noloc' }
# NOTE: tbl_fields dont contain PRIMAY key or SERIAL columns, like *id* in wpp_uprecsinfo.
tbl_field = { 'wpp_clusteridaps':('clusterid', 'keyaps', 'seq'),
                      'wpp_cfps':('clusterid', 'lat', 'lon', 'height', 'rsss', 'cfps_time'),
                'wpp_uprecsinfo':('spid','servid','time','imsi','imei','useragent',
                                  'mcc','mnc','lac','cellid','cellrss',
                                  'lat','lon','height','wlanidentifier','wlanmatcher',
                                  'ver_uprecs'),
              'wpp_uprecs_noloc':('spid','servid','time','imsi','imei','useragent',
                                  'mcc','mnc','lac','cellid','cellrss',
                                  'lat','lon','height','wlanidentifier','wlanmatcher',
                                  'ver_uprecs'),
                        'tsttbl':('clusterid', 'keyaps', 'seq') }
tbl_idx =   { 'wpp_clusteridaps':('clusterid','keyaps'), #{table_name:{'field_name'}}
                      'wpp_cfps':('clusterid',),
                'wpp_uprecsinfo':('ver_uprecs',),
              'wpp_uprecs_noloc':('ver_uprecs',),
                 'wpp_uprecsver':(),
                        'tsttbl':('clusterid',)}
tbl_files = { 'wpp_clusteridaps':'tbl/cidaps.tbl', 
                      'wpp_cfps':'tbl/cfprints.tbl',
                'wpp_uprecsinfo':'tbl/uprecs.tbl',
              'wpp_uprecs_noloc':'tbl/uprecs_noloc.tbl',
                 'wpp_uprecsver':'tbl/uprecsver.tbl',
                        'cidaps':'tbl/cidaps.tbl',
                          'cfps':'tbl/cfprints.tbl',
                        'tsttbl':'tbl/tsttbl.tbl' }
tbl_forms = { 'oracle':{
                'wpp_clusteridaps':""" (  
                     clusterid INT NOT NULL, 
                        keyaps VARCHAR2(71) NOT NULL,
                           seq INT NOT NULL)""", 
                'wpp_cfps':""" (  
                     clusterid INT NOT NULL,
                           lat NUMBER(9,6) NOT NULL,
                           lon NUMBER(9,6) NOT NULL,
                        height NUMBER(5,1) DEFAULT 0,
                          rsss VARCHAR2(100) NOT NULL,
                     cfps_time VARCHAR2(20))""",
                'wpp_uprecsinfo':""" (  
                            id INT PRIMARY KEY,	
                          spid INT,
                        servid INT,
                          time VARCHAR(20),
                          imsi VARCHAR(20),
                          imei VARCHAR(20),
                     useragent VARCHAR(300),
                           mcc INT,
                           mnc INT,
                           lac INT,
                        cellid INT,
                       cellrss VARCHAR(5),
                           lat NUMERIC(9,6),
                           lon NUMERIC(9,6),
                        height NUMERIC(5,1),
                wlanidentifier VARCHAR(1024),
                   wlanmatcher VARCHAR(255))""",
                'tsttbl':"""(
                     clusterid INT, 
                        keyaps VARCHAR2(71) NOT NULL,
                           seq INT NOT NULL)""" },
              'postgresql':{
                'wpp_clusteridaps':"""(
                     clusterid INT NOT NULL, 
                        keyaps VARCHAR(360) NOT NULL,
                           seq INT NOT NULL)""", 
                'wpp_cfps':""" (
                     clusterid INT NOT NULL,
                           lat NUMERIC(9,6) NOT NULL,
                           lon NUMERIC(9,6) NOT NULL,
                        height NUMERIC(5,1) DEFAULT 0,
                          rsss VARCHAR(100) NOT NULL,
                     cfps_time VARCHAR(20))""",
                'wpp_uprecsver':""" (
                    ver_uprecs INT DEFAULT 0)""",
                'wpp_uprecsinfo':""" (
                          spid INT,
                        servid INT,
                          time VARCHAR(20),
                          imsi VARCHAR(20),
                          imei VARCHAR(20),
                     useragent VARCHAR(300),
                           mcc INT DEFAULT 0,
                           mnc INT DEFAULT 0,
                           lac INT DEFAULT 0,
                        cellid INT DEFAULT 0,
                       cellrss VARCHAR(5),
                           lat NUMERIC(9,6) DEFAULT 0,
                           lon NUMERIC(9,6) DEFAULT 0,
                        height NUMERIC(5,1) DEFAULT 0,
                wlanidentifier VARCHAR(1024),
                   wlanmatcher VARCHAR(255),
                    ver_uprecs INT DEFAULT 0)""",
                'wpp_uprecs_noloc':""" (
                          spid INT,
                        servid INT,
                          time VARCHAR(20),
                          imsi VARCHAR(20),
                          imei VARCHAR(20),
                     useragent VARCHAR(300),
                           mcc INT DEFAULT 0,
                           mnc INT DEFAULT 0,
                           lac INT DEFAULT 0,
                        cellid INT DEFAULT 0,
                       cellrss VARCHAR(5),
                           lat NUMERIC(9,6) DEFAULT 0,
                           lon NUMERIC(9,6) DEFAULT 0,
                        height NUMERIC(5,1) DEFAULT 0,
                wlanidentifier VARCHAR(1024),
                   wlanmatcher VARCHAR(255),
                    ver_uprecs INT DEFAULT 0)""",
                'tsttbl':"""(
                     clusterid INT, 
                        keyaps VARCHAR2(71) NOT NULL,
                           seq INT NOT NULL)""" }}
# SQL statements.
sqls = { 'SQL_SELECT' : "SELECT %s FROM %s",
         'SQL_UPDATE' : "UPDATE %s SET %s = %s",
         'SQL_DELETE' : "DELETE FROM %s WHERE %s",
         'SQL_DROPTB' : "DROP TABLE %s PURGE",
         'SQL_INSERT' : "INSERT INTO %s %s VALUES %s",
  'SQL_INSERT_SELECT' : "INSERT INTO %s SELECT %s FROM %s",
        'SQL_TRUNCTB' : "TRUNCATE TABLE %s",
        'SQL_DROP_MY' : "DROP TABLE IF EXISTS %s",
       'SQL_DROP_IDX' : "DROP INDEX %s",
    'SQL_DROP_IDX_IE' : "DROP INDEX IF EXISTS %s",
       'SQL_CREATETB' : "CREATE TABLE %s %s",
      'SQL_CREATEIDX' : "CREATE INDEX %s ON %s(%s)",
     'SQL_CREATEUIDX' : "CREATE UNIQUE INDEX %s ON %s(%s)",
    'SQL_CREATETB_MY' : "CREATE TABLE IF NOT EXISTS %s %s",
       'SQL_CSVIN_MY' : """
                        LOAD DATA LOCAL INFILE "%s" INTO TABLE %s 
                        FIELDS TERMINATED BY ',' 
                        LINES TERMINATED BY '\\n' 
                        %s""" }

# Text colors
termtxtcolors = {
        'red':'\033[91m%s\033[0m',
      'green':'\033[92m%s\033[0m',
     'yellow':'\033[93m%s\033[0m',
       'blue':'\033[94m%s\033[0m',
     'purple':'\033[95m%s\033[0m'
}
# String length of 179 and 149 chars are used for each intersection set to have 
# at most INTERSET APs, which should be enough for classification, very ugly though.
#dt_rmp_nocluster = {'names':('spid','lat','lon','macs','rsss'), 
#                  'formats':('i4','f4','f4','S179','S149')}
WLAN_FAKE = {
        1: #home
            [ ['00:25:86:23:A4:48', '-86'], ['00:24:01:FE:0F:20', '-90'], 
              ['00:0B:6B:3C:75:34', '-89'] ],
        2: #home-only 1 visible
            [ ['00:0B:6B:3C:75:34', '-89'] ],
        3: #cmri-only 1 visible
            [ ['00:15:70:9E:91:60', '-53'] ],
        4: #cmri-fail
            [ ['00:15:70:9F:7D:88', '-82'], ['00:15:70:9F:7D:89', '-77'],
              ['00:15:70:9F:7D:8A', '-77'], ['00:23:89:3C:BD:F2', '-81'],
              ['00:11:B5:FD:8B:6D', '-81'], ['00:23:89:3C:BE:10', '-70'],
              ['00:23:89:3C:BE:11', '-70'], ['00:23:89:3C:BE:13', '-71'],
              ['00:15:70:9E:91:62', '-72'], ['00:15:70:9E:91:60', '-49'],
              ['00:23:89:3C:BD:32', '-75'], ['00:15:70:9E:91:61', '-50'],
              ['00:23:89:3C:BE:12', '-75'], ['00:23:89:3C:BD:33', '-76'],
              ['00:14:BF:1B:A5:48', '-79'], ['00:15:70:9E:6C:6D', '-68'],
              ['00:15:70:9E:6C:6C', '-68'], ['00:15:70:9E:6C:6E', '-68'],
              ['00:23:89:3C:BD:30', '-75'], ['00:23:89:3C:BD:31', '-75'],
              ['00:23:89:3C:BC:90', '-79'], ['00:23:89:3C:BC:93', '-75'],
              ['00:11:B5:FE:8B:6D', '-88'], ['00:23:89:3C:BC:91', '-80'],
              ['00:23:89:3C:BC:92', '-81'], ['00:23:89:3C:BD:F1', '-80']],
        5: #cmri-ok-part
            [ ['00:15:70:9F:7D:8A', '-76'], ['00:15:70:9F:7D:88', '-77'],
              ['00:15:70:9F:7D:89', '-80'], ['00:11:B5:FD:8B:6D', '-79'],
              ['00:23:89:3C:BC:90', '-75'], ['00:23:89:3C:BC:91', '-76'],
              ['00:23:89:3C:BC:92', '-76'], ['00:23:89:3C:BC:93', '-75'],
              ['00:23:89:3C:BE:12', '-73'], ['00:23:89:3C:BE:10', '-75'],
              ['00:23:89:3C:BE:11', '-69'], ['00:15:70:9E:91:61', '-63'],
              ['00:23:89:3C:BE:13', '-71'], ['00:15:70:9E:91:62', '-61'],
              ['00:15:70:9E:91:60', '-62'], ['00:14:BF:1B:A5:48', '-81'],
              ['00:23:89:3C:BD:33', '-73'], ['00:15:70:9E:6C:6C', '-67'],
              ['00:15:70:9E:6C:6D', '-68'], ['00:15:70:9E:6C:6E', '-67']],
        6: #cmri-ok-full
            [ ['00:11:B5:FD:8B:6D', '-69'], ['00:15:70:9E:91:60', '-52'], 
              ['00:15:70:9E:91:61', '-53'], ['00:15:70:9F:73:64', '-78'], 
              ['00:15:70:9F:73:66', '-75'], ['00:15:70:9E:91:62', '-55'],
              ['00:23:89:3C:BE:10', '-74'], ['00:23:89:3C:BE:11', '-78'], 
              ['00:23:89:3C:BE:12', '-78'], ['00:11:B5:FE:8B:6D', '-80'], 
              ['00:15:70:9E:6C:6C', '-65'], ['00:15:70:9E:6C:6D', '-60'],
              ['00:15:70:9E:6C:6E', '-70'], ['00:15:70:9F:76:E0', '-81'], 
              ['00:15:70:9F:7D:88', '-76'], ['00:15:70:9F:73:65', '-76'], 
              ['00:23:89:3C:BD:32', '-75'], ['00:23:89:3C:BD:30', '-78'],
              ['02:1F:3B:00:01:52', '-76'] ],
        7: #cmri-square-fail
            [ ['00:16:16:1F:14:E0', '-49'], ['00:16:16:1E:EB:60', '-78'] ],
        8: #hq-fail
            [ ['00:60:B3:C9:61:27', '-63'], ['00:16:16:1E:B9:80', '-64'],
              ['00:1A:70:FB:B8:7F', '-65'], ['00:17:7B:0F:16:D9', '-66'] ],
        9: #hq-fail
            [ ['00:60:B3:C9:61:27', '-61'], ['00:1B:54:25:86:40', '-64'],
              ['00:17:7B:0F:16:D8', '-66'], ['00:17:7B:0F:16:D9', '-65'] ],
        10:#hq-ok
            [ ['00:60:B3:C9:61:27', '-64'], ['00:1E:E3:E0:69:40', '-64'],
              ['00:17:7B:0F:16:D8', '-66'], ['00:16:16:1E:B9:80', '-65'] ],
        11:#hq-fail
            [ ['00:60:B3:C9:61:27', '-66'], ['00:16:16:1E:82:20', '-67'],
              ['00:17:7B:0F:16:D8', '-67'], ['00:16:16:1F:24:A0', '-67'] ],
        12:#hq-fail
            [ ['00:60:B3:C9:61:27', '-63'], ['00:16:16:1E:B9:80', '-66'],
              ['00:17:7B:0F:16:D8', '-66'], ['00:16:16:1E:78:C0', '-69'] ],
        13:#hq-fail
            [ ['00:60:B3:C9:61:27', '-65'], ['00:17:7B:0F:16:D9', '-67'],
              ['00:17:7B:0F:16:DA', '-67'], ['00:1B:53:6C:E7:B0', '-67'] ],
        14:#hq-fail
            [ ['00:60:B3:C9:61:27', '-64'], ['00:17:7B:0F:16:D9', '-69'],
              ['00:16:16:1E:B9:80', '-65'], ['00:16:16:1F:30:60', '-68'] ],
        15:#hq-fail
            [ ['00:60:B3:C9:61:27', '-65'], ['00:17:7B:0F:16:D9', '-67'],
              ['00:1B:54:25:86:40', '-64'], ['00:16:16:1F:30:60', '-66'] ],
        16:#hq-fail
            [ ['00:60:B3:C9:61:27', '-65'], ['00:17:7B:0F:16:D9', '-66'],
              ['00:1B:54:25:86:40', '-66'], ['00:17:7B:0F:16:D8', '-66'] ],
        17:#hq-fail-dknn_0_and_1
            [ ['00:1E:E3:E0:69:40', '-66'], ['00:1D:7E:51:E0:8D', '-69'],
              ['00:16:16:1E:82:20', '-69'], ['00:17:7B:0F:16:D8', '-69'] ],
        18:#hq-square-interpolated-between-1313-and-902
            [ ['00:23:89:5F:D8:A1', '-71'], ['00:15:70:D0:52:60', '-71'],
              ['00:15:70:D0:52:61', '-73'], ['00:23:89:5C:9E:D0', '-73'] ],
        19:#hq-square
            [ ['00:17:7B:FC:34:70', '-60'], ['00:15:70:D0:52:62', '-65'],
              ['00:23:89:3C:BD:12', '-67'], ['00:15:70:D0:52:60', '-67'] ],
        20:#hq-square
            [ ['00:17:7B:FC:34:70', '-61'], ['00:15:70:D0:52:62', '-68'],
              ['00:15:70:D0:52:61', '-68'], ['00:23:89:3C:BD:13', '-69'] ],
        21:#bigerr-wpp: 905.11, signal too weak!!
            [ ['00:19:E0:E3:85:D0','-92'] ],
        22:#bigerr-wpp: 162.56
            [ ['00:17:7B:0F:85:F0','-68'] ],
        23:#issue 24: mismatch between queried macs/rsss.
            [ ['00:17:7B:0F:0F:08', '-73'], ['00:1F:A4:05:AF:3E', '-78'],
              ['00:24:01:25:0B:CC', '-79'], ['74:EA:3A:4C:B8:58', '-81'] ],
        24:#issue 25: K nearest FPs, which should be K nearest distances.
            [ ['52:C3:17:AE:71:51', '-91'] ],
        25:
            [ ['00:21:91:1D:C0:D4', '-90'], ['00:25:86:4D:B4:C4', '-90'] ],
        26:
            [ ['00:17:7B:0F:0F:58', '-85'], ['00:27:19:52:B5:20', '-83'] ],
        27:
            [ ['00:21:91:1D:C1:06', '-84'] ],
        28: # nan
            [ ['00:23:69:D8:B0:80', '-81'] ],
        29: # nan
            [ ['00:21:27:50:99:AC', '-82'] ],
        30: # nan
            [ ['00:27:19:64:DE:78', '-85'] ],
        31: # 
            [ ['00:0F:3D:0B:51:28', '-89'], ['00:1F:A3:B6:7B:C0', '-80'],
              ['00:21:27:50:99:AC', '-77'], ['00:27:19:9E:48:A0', '-78'],
              ['D8:5D:4C:63:FA:A6', '-88'] ],
        32: # bigerr 135.20
            [ ['00:21:91:1D:C1:06', '-84'] ],
        33: # bigerr 1581.43
            [ ['00:b0:0c:4b:75:c0', '-82'], ['00:25:86:37:19:2e', '-84'],
              ['00:90:4c:7e:00:64', '-85'], ['00:19:e0:e0:3f:7c', '-85'] ],
        34: # crossover of bigerr 1581.43
            [ ['00:b0:0c:4b:75:c0', '-82'], ['00:25:86:37:19:2e', '-84'],
              ['00:90:4c:7e:00:64', '-85'], ['00:17:7b:fc:34:28', '-87'] ],
        35: # bigerr 791
            [ ['00:17:7b:0f:0f:58', '-78'], ['00:b0:0c:0f:3e:98', '-86'],
              ['00:17:7b:0f:7a:b0', '-90'], ['1c:af:f7:a7:ba:b4', '-91'] ],
            
}
icon_types = { 'on': [ '"encrypton"',  '/kml/icons/encrypton.png'],
              'off': [ '"encryptoff"', '/kml/icons/encryptoff.png'],
           'reddot': [ '"reddot"',     '/kml/icons/reddot.png'],
          'bluedot': [ '"bluedot"',    '/kml/icons/bluedot.png'],
        'yellowdot': [ '"yellowdot"',  '/kml/icons/yellowdot.png'],
             'wifi': [ '"wifi"',       '/kml/icons/wifi.png'],
        'dotshadow': [ '"dotshadow"',  '/kml/icons/dotshadow.png'],
}

props_jpg = {'term':'jpeg', # MUST be recognized by Gnuplot.
         'outfname':'cdf.jpg',
             'font':'"/usr/share/fonts/truetype/arphic/gbsn00lp.ttf, 14"',
             'size':'', # default: 1,1
            'title':'误差累积函数',
           'xlabel':'误差/米',
           'ylabel':'概率',
           'legend':'',
              'key':'right bottom',
           'xrange':[0,100],
           'yrange':[0,1],
            'xtics':'nomirror 10',
            'ytics':'nomirror .05',
             'grid':'x y', # default: off
           'border':3,
             'with':'lp pt 3 lc 1'}
props_mp = { 'term':'mp latex', # MUST be recognized by Gnuplot.
         'outfname':'cdf.mp',
             'font':'"Romans" 7',
             'size':'.8, .8', # default: 1,1
            'title':'CDF',
           'xlabel':'error/m',
           'ylabel':'probability',
           'legend':'',
              'key':'right bottom',
           'xrange':[0,100],
           'yrange':[0,1],
            'xtics':'nomirror 10',
            'ytics':'nomirror .05',
             'grid':'x y', # default: off
           'border':3,
             'with':'lp pt 4'}
